### PR TITLE
Deploy consumer 

### PR DESCRIPTION
**Pull request by DevAseel**
**Summary of change:**
- Added a new micro-service, deploy consumer, which uses deployFetcher, deployParser, and deploySaver to consume a deploy. 

**List of tasks:** 

- [x] create class
- [x] create mocks (MockDeployFetcher, MockDeployParser, MockDeploySaver)
- [x] create deploy fetcher unit test.
- [x] create deploy consumer unit test.
- [x] create deploy consumer integration test
- [x] update user guide
- [x] update changelog
- [x] usage test: Last check for the performance of the service where the developer imitates the user experience by importing the services in a similar way to the npm package.

**Steps for testing**
- Make sure your local git is up to date
```bash
git pull
```

- Checkout the branch `feature/deploy-consumer`
```bash
git checkout feature/deploy-consumer
```

- Run the test across the test-net
```bash
npm run test
```

- Confirm deploy-consumer test has passed.
- Run the docs using live server and make sure it have been updated. 

**Proof of Success**

Screenshot 2022-12-11 at 9 24 03 PM

### CODE REVIEW

File: src/services/consumers/DeployConsumer.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    const deployParserResult: DeployParserResult = this.deployParser.apply(result.fields);
+
+    if (!deployParserResult.success) {
+      const message = `failed to parse deploy with hash of ${hash}`;
+      this.error(message);
+      return { success: false, message, hash };
+    }
+
```
Comment (jjfluid) [inline src/services/consumers/DeployConsumer.ts:60]: I think there needs to be a step right here - if the contractHash field is undefined in the result, we will need to call another intermediate service that can fetch/determine the contract_hash, whether its from blockchain requests or API calls to Caspernet. idk

Comment (DevAseel) [reply src/services/consumers/DeployConsumer.ts:60]: @jjfluid I will try to research and find away for doing that. There has to be a more efficient way of doing it.

**Comment by DevAseel** [comment]
This service is now complete, will add [the proposed](https://github.com/fluidefi/fluidefi-caspernet-aggregator-tools/pull/46) changes here to avoid conflicts.

**Comment by DevAseel** [comment]
@specialistvlad This PR is now ready for review. It contains all of the changes in [#45](https://github.com/fluidefi/fluidefi-caspernet-aggregator-tools/pull/45) & [#46](https://github.com/fluidefi/fluidefi-caspernet-aggregator-tools/pull/46). The two PRs should be closed, and this PR should be merged.

File: src/services/datastore/DeploySaver.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        // save block to data store
+        await repository.save(deploy);
+
+        return {
+          success: true,
+          hash: fields.deployHash,
+        };
+      } catch (e: any) {
+        // catch errors with uniqueness constraint violations
+        if (e.toString().includes('unique')) {
```
Comment (specialistvlad) [inline src/services/datastore/DeploySaver.ts:57]: Can we have here something like this `if (e instanceof ErrorClass)` @DevAseel

File: src/services/datastore/DeploySaver.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+            message: warningMessage,
+          };
+        }
+
+        // return with failure for other errors
+        const errorMessage = `failed to save deploy with fields: ${fields ? JSON.stringify(fields) : undefined}`;
+        this.error(errorMessage, e);
+        return { success: false, error: e, message: errorMessage };
+      }
+    } catch (e: any) {
```
Comment (specialistvlad) [inline src/services/datastore/DeploySaver.ts:73]:
```suggestion
    } catch (e: Error) {
```

File: src/services/datastore/DeploySaver.ts (hunk 2 above)
Comment (specialistvlad) [reply src/services/datastore/DeploySaver.ts:57]: https://fettblog.eu/typescript-typing-catch-clauses/

File: test/mocks/MockDeployFetcher.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  constructor() {
+    this.deploys = [];
+  }
+
+  injectDeploys(deploys: any): void {
+    this.deploys = deploys || [];
+  }
+
+  async apply(hash: string, blockNumber: number): Promise<DeployFetcherResult> {
+    if (!!this.deploys) {
```
Comment (specialistvlad) [inline test/mocks/MockDeployFetcher.ts:16]:
```
if (!!this.deploys) { 
 return { success: false, error: 'Deploy not found', hash, blockNumber };
}

... rest of the code
```

Does it make sense to reduce depth of code?

File: src/services/datastore/DeploySaver.ts (hunk 2 above)
Comment (specialistvlad) [reply src/services/datastore/DeploySaver.ts:57]: @jjordan-quantum we are took this code from some file near as example. So the question is: is it ok to have the error checking like this for now?

Comment (specialistvlad) [reply src/services/datastore/DeploySaver.ts:57]: Task created CS-72

Comment (jjfluid) [reply src/services/datastore/DeploySaver.ts:57]: find the uniqueness constraint violation error class for typeorm is a good idea

File: src/services/consumers/DeployConsumer.ts (hunk 1 above)
Comment (jjfluid) [reply src/services/consumers/DeployConsumer.ts:61]: You can create the logic to check this in the parser with additional info in the return value maybe, then create an interface / stub service for getting the missing data.

If you want to tackle this now, I will hold off on approving the PR

File: test/mocks/MockDeployFetcher.ts (hunk 4 above)
Comment (DevAseel) [reply test/mocks/MockDeployFetcher.ts]: Done

File: src/services/consumers/DeployConsumer.ts (hunk 1 above)
Comment (jjfluid) [reply src/services/consumers/DeployConsumer.ts:61]: sounds good

**Review by jjfluid — APPROVED** [review]

**Pull request merged**